File: java/revenj-core/src/main/java/org/revenj/database/postgres/converters/ByteaConverter.java
package org.revenj.database.postgres.converters;

import org.revenj.database.postgres.PostgresBuffer;
import org.revenj.database.postgres.PostgresWriter;
import org.revenj.database.postgres.PostgresReader;

import java.util.ArrayList;
import java.util.List;

public abstract class ByteaConverter {

	private static final char[] CharMap;
	private static final int[] CharLookup;

	static {
		CharMap = "0123456789abcdef".toCharArray();
		CharLookup = new int['f' + 1];
		for (int i = 0; i < CharMap.length; i++) {
			CharLookup[CharMap[i]] = i;
		}
	}

	private static final byte[] EMPTY_BYTES = new byte[0];
	private static final char[] XX = "\\x".toCharArray();

	public static void serializeURI(PostgresBuffer sw, byte[] value) {
		sw.addToBuffer(XX);
		for (byte b : value) {
			sw.addToBuffer(CharMap[(b >> 4) & 0xf]);
			sw.addToBuffer(CharMap[b & 0xf]);
		}
	}

	private static byte[] toArray(List<Byte> list) {
		byte[] result = new byte[list.size()];
		for (int i = 0; i < list.size(); i++) {
			result[i] = list.get(i);
		}
		return result;
	}

	public static byte[] parse(PostgresReader reader, int context) {
		int cur = reader.read();
		if (cur == ',' || cur == ')') {
			return null;
		}
		int len = context + (context << 1);
		if (len == 0) {
			len = 1;
		}
		cur = reader.read(len + 1);
		List<Byte> list = new ArrayList<>(1024);
		while (cur != -1 && cur != '\\' && cur != '"') {
			list.add((byte) ((CharLookup[cur] << 4) + CharLookup[reader.read()]));
			cur = reader.read();
		}
		reader.read(context);
		return toArray(list);
	}

	public static List<byte[]> parseCollection(PostgresReader reader, int context, boolean allowNulls) {
		int cur = reader.read();
		if (cur == ',' || cur == ')') {
			return null;
		}
		boolean escaped = cur != '{';
		if (escaped) {
			reader.read(context);
		}
		int innerContext = context << 1;
		int skipInner = innerContext + (innerContext << 1);
		List<byte[]> list = new ArrayList<>();
		cur = reader.peek();
		if (cur == '}') {
			reader.read();
		}
		byte[] emptyColl = allowNulls ? null : EMPTY_BYTES;
		while (cur != -1 && cur != '}') {
			cur = reader.read();
			if (cur == 'N') {
				list.add(emptyColl);
				cur = reader.read(4);
			} else {
				reader.read(skipInner);
				List<Byte> item = new ArrayList<>(1024);
				cur = reader.read();
				while (cur != -1 && cur != '"' && cur != '\\') {
					item.add((byte) ((CharLookup[cur] << 4) + CharLookup[reader.read()]));
					cur = reader.read();
				}
				cur = reader.read(innerContext);
				list.add(toArray(item));
			}
		}
		if (escaped) {
			reader.read(context + 1);
		} else {
			reader.read();
		}
		return list;
	}

	public static PostgresTuple toTuple(byte[] value) {
		return value != null ? new ByteTuple(value) : null;
	}

	private static class ByteTuple extends PostgresTuple {
		private final byte[] value;

		ByteTuple(byte[] value) {
			this.value = value;
		}

		public boolean mustEscapeRecord() {
			return true;
		}

		public boolean mustEscapeArray() {
			return true;
		}

		private void buildArray(PostgresWriter sw) {
			for (byte b : value) {
				sw.write(CharMap[(b >> 4) & 0xf]);
				sw.write(CharMap[b & 0xf]);
			}
		}

		public void buildTuple(PostgresWriter sw, boolean quote) {
			if (quote) {
				sw.write('\'');
				insertRecord(sw, "", null);
				sw.write('\'');
			} else insertRecord(sw, "", null);
		}

		public void insertRecord(PostgresWriter sw, String escaping, Mapping mappings) {
			String pref = buildSlashEscape(escaping.length());
			if (mappings != null) {
				for (int x = 0; x < pref.length(); x++)
					mappings.map(sw, pref.charAt(x));
			} else sw.write(pref);
			sw.write('x');
			buildArray(sw);
		}

		public void insertArray(PostgresWriter sw, String escaping, Mapping mappings) {
			//TODO this is wrong
			insertRecord(sw, escaping, mappings);
		}
	}
}

File: scala/revenj-core/src/main/scala/net/revenj/database/postgres/converters/ByteaConverter.scala
package net.revenj.database.postgres.converters

import net.revenj.database.postgres.{PostgresBuffer, PostgresReader, PostgresWriter}

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer

object ByteaConverter extends Converter[Array[Byte]] {

  private val charMap = "0123456789abcdef".toCharArray
  private val charLookup = {
    val result = new Array[Int]('f' + 1)
    var i = 0
    while (i < charMap.length) {
      result(charMap(i).toInt) = i
      i += 1
    }
    result
  }

  private val EMPTY_BYTES = new Array[Byte](0)
  private val XX = "\\x".toCharArray

  override def serializeURI(sw: PostgresBuffer, value: Array[Byte]): Unit = {
    sw.addToBuffer(XX)
    var i = 0
    while (i < value.length) {
      val b = value(i)
      sw.addToBuffer(charMap((b >> 4) & 0xf))
      sw.addToBuffer(charMap(b & 0xf))
      i += 1
    }
  }

  val dbName = "bytea"

  def default() = EMPTY_BYTES

  override def parseRaw(reader: PostgresReader, start: Int, context: Int): Array[Byte] = {
    val len = if (context == 0) 1 else context + (context << 1)
    parseBytea(reader, len, context)
  }

  private def parseBytea(reader: PostgresReader, len: Int, context: Int): Array[Byte] = {
    var cur = reader.read(len + 1)
    val builder = mutable.ArrayBuilder.make[Byte]()
    builder.sizeHint(1024)
    while (cur != -1 && cur != '\\' && cur != '"') {
      builder += ((charLookup(cur) << 4) + charLookup(reader.read())).toByte
      cur = reader.read()
    }
    reader.read(context)
    builder.result()
  }

  override def parseCollectionOption(reader: PostgresReader, context: Int): Option[ArrayBuffer[Array[Byte]]] = {
    var cur = reader.read()
    if (cur == ',' || cur == ')') {
      None
    } else {
      val escaped = cur != '{'
      if (escaped) {
        reader.read(context)
      }
      val innerContext = context << 1
      val skipInner = innerContext + (innerContext << 1)
      cur = reader.peek
      if (cur == '}') {
        reader.read()
      }
      val list = ArrayBuffer.newBuilder[Array[Byte]]
      while (cur != -1 && cur != '}') {
        cur = reader.read()
        if (cur == 'N') {
          list += EMPTY_BYTES
          reader.read(4)
        } else {
          list += parseBytea(reader, skipInner, innerContext)
        }
        cur = reader.last
      }
      if (escaped) {
        reader.read(context + 1)
      } else {
        reader.read()
      }
      Some(list.result())
    }
  }

  override def parseNullableCollectionOption(reader: PostgresReader, context: Int): Option[ArrayBuffer[Option[Array[Byte]]]] = {
    var cur = reader.read()
    if (cur == ',' || cur == ')') {
      None
    } else {
      val escaped = cur != '{'
      if (escaped) {
        reader.read(context)
      }
      val innerContext = context << 1
      val skipInner = innerContext + (innerContext << 1)
      cur = reader.peek
      if (cur == '}') {
        reader.read()
      }
      val list = ArrayBuffer.newBuilder[Option[Array[Byte]]]
      while (cur != -1 && cur != '}') {
        cur = reader.read()
        if (cur == 'N') {
          list += None
          reader.read(4)
        } else {
          list += Some(parseBytea(reader, skipInner, innerContext))
        }
        cur = reader.last
      }
      if (escaped) {
        reader.read(context + 1)
      } else {
        reader.read()
      }
      Some(list.result())
    }
  }

  override def parseCollectionItem(reader: PostgresReader, context: Int): Array[Byte] = {
    val innerContext = context + (context << 1)
    parseBytea(reader, innerContext, context)
  }

  override def parseNullableCollectionItem(reader: PostgresReader, context: Int): Option[Array[Byte]] = {
    val cur = reader.read()
    if (cur == 'N') {
      reader.read(4)
      None
    } else {
      val innerContext = context + (context << 1)
      Some(parseBytea(reader, innerContext, context))
    }
  }

  def toTuple(value: Array[Byte]): PostgresTuple = {
    new ByteaTuple(value)
  }

  private class ByteaTuple(val value: Array[Byte]) extends PostgresTuple {

    val mustEscapeRecord = true

    val mustEscapeArray = true

    private def buildArray(sw: PostgresWriter): Unit = {
      var i = 0
      while (i < value.length) {
        val b = value(i)
        sw.addToBuffer(charMap((b >> 4) & 0xf))
        sw.addToBuffer(charMap(b & 0xf))
        i += 1
      }
    }

    override def buildTuple(sw: PostgresWriter, quote: Boolean): Unit = {
      if (quote) {
        sw.write('\'')
        insertRecord(sw, "", None)
        sw.write('\'')
      }
      else insertRecord(sw, "", None)
    }

    def insertRecord(sw: PostgresWriter, escaping: String, mappings: Option[(PostgresWriter, Char) => Unit]): Unit = {
      lazy val pref = PostgresTuple.buildSlashEscape(escaping.length)
      mappings match {
        case Some(m) =>
          var x = 0
          while (x < pref.length) {
            m(sw, pref.charAt(x))
            x += 1
          }
        case _ =>
          sw.write(pref)
      }
      sw.write('x')
      buildArray(sw)
    }

    override def insertArray(sw: PostgresWriter, escaping: String, mappings: Option[(PostgresWriter, Char) => Unit]): Unit = {
      insertRecord(sw, escaping, mappings)
    }
  }

}
